### PR TITLE
[NET-0008] Add authentication and permissions

### LINKED ISSUE #8: NET-0008: Implement authentication functionality
**DoD**
- [x] Implemented signup, login, and logout endpoint.
- [x] Utility functions should be implemented, which do role verification, and check whether a user is logged in.
- [x] GET, POST, and DELETE endpoints should be implemented for /users. Functionalities for each involve basic Read, Create, and Delete functions respectively.
- [x] GET, PUT, and DELETE endpoints should be implemented for /users/:userId. Functionalities for each involve basic Read, Update, and Delete functions respectively, for specific users identified with userId.
- [x] Implement the ff. roles: user, operator, manager, and admin
- [x] Admin users cannot be created, updated, or deleted. There can only be one admin.

**Endpoint**
/signup (POST)
/login (POST)
/logout (GET)
/users
- GET (manager and admin)
- POST (admin)
- DELETE (admin)
/users/:userId
- GET (manager and admin)
- PUT (manager and admin)
- DELETE (manager and admin)

**Additional context**
Related Story:
As an owner, I want to have the option to allow only selected users to access certain data and functionalities, to prevent unwanted data access and modifications.

### PR DESCRIPTION
**Pull request by RayAdrian**
## Description
Resolves #8 

**Changelog:**
- [x] Implemented signup and login endpoints.
- [x] Utility functions should be implemented, which do role verification, and check whether a user is logged in.
- [x] GET, POST, and DELETE endpoints should be implemented for /users. Functionalities for each involve basic Read, Create, and Delete functions respectively.
- [x] GET, PUT, and DELETE endpoints should be implemented for /users/:userId. Functionalities for each involve basic Read, Update, and Delete functions respectively, for specific users identified with userId.
- [x] Implement the ff. roles: user, operator, manager, and admin

**How to Test:**
1. See swagger for schemas
2. On postman, create a user using /users/signup endpoints
3. Login using /users/login
4. Place the token in Authorization -> Token for succeeding API calls

### CODE REVIEW

File: routes/cors.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+const express = require('express');
+const cors = require('cors');
+// eslint-disable-next-line no-unused-vars
+const app = express();
```
Comment (allendomingo) [inline routes/cors.js]: If this is unused, can this be removed?

File: routes/users.js [hunk 2]
```diff
@@ -1,12 +1,232 @@
...
+  userController.getUsers(req.query)
+  .then(users => {
+    res.statusCode = 200;
+    res.setHeader('Content-Type', 'application/json');
+    res.json(users);
+  }, err => next(err))
+  .catch(err => next(err));
+});
+
+router.post('/', (_, res) => {
```
Comment (allendomingo) [inline routes/users.js:34]: Inconsistent format for callback functions. I suggest sticking with callback functions for this,

Comment (RayAdrian) [reply routes/users.js:34]: You mean arrow functions?

Comment (allendomingo) [reply routes/users.js:34]: yes

File: app.js [hunk 3]
```diff
@@ -40,12 +41,19 @@ app.use(express.json());
...
 
 // swagger documentation
 if (process.env.NODE_ENV === 'development') {
   const swaggerUi = require('swagger-ui-express');
-  const swaggerFile = require('./bin/swagger-output.json');
-  app.use('/api-docs', swaggerUi.serve, swaggerUi.setup(swaggerFile));
+
+  // Implement these in a try catch block since swagger-output is not always available (gitignored)
+  try {
+    const swaggerFile = require('./bin/swagger-output.json');
```
Comment (RayAdrian) [inline app.js]: Lint CI is complaining about this but it is gitignored. Please advise @allendomingo

**Review by allendomingo — APPROVED** [review]
Looks good to me

**Pull request merged**